### ISSUE TITLE
l5d-dst-override header not working

### ISSUE THREAD

**Issue by oncipriani** [linkerd/linkerd2#9642]
### What is the issue?

Per route metrics are only working when we add the annotation `nginx.ingress.kubernetes.io/upstream-vhost: $service_name.$namespace.svc.cluster.local` to the application's Ingress.

Unfortunately, this annotation changes the `Host` header forwarded to the application, which breaks the URLs that it generates and displays to the end users.

This should be an easy fix, according to the documentation for [Service Profiles](https://linkerd.io/2.12/tasks/setting-up-service-profiles/):

> The destination service for a request is computed by selecting the value of the first header to exist of, l5d-dst-override, :authority, and Host.

So, we **replaced** the annotation above with the following:

```yaml
nginx.ingress.kubernetes.io/configuration-snippet: |
  proxy_set_header l5d-dst-override "$service_name.$namespace.svc.cluster.local";
```

It did not work. Per route metrics stopped being collected, even though we validated that the `l5d-dst-override` header was being forwarded to the application with the correct value (the value matches the name of the Service Profile exactly).

### How can it be reproduced?

1. Configure an application to collect per-route metrics;
2. Make sure the `Host` header matches the Service Profile name;
3. Notice that per route metrics are collected;
4. Change the `Host` header value to something else;
5. Add the `l5d-dst-override` header and make sure its value matches the Service Profile name;
6. Notice that per route metrics are no longer working.

### Logs, error output, etc

The output of `linkerd viz tap deploy/webapp -o wide` correctly includes `rt_route` when the `Host` header matches the Service Profile name, where it is missing in all other cases.

### output of `linkerd check -o short`

```text
Status check results are √
```

### Environment

- Kubernetes version: 1.23
- Cluster environment: RKE
- Host OS: Rocky Linux 8.6
- Linkerd version: stable-2.12.1

### Possible solution

It is possible that the `Host` header is being seen by Linkerd before the `l5d-dst-override` header and in this case the `l5d-dst-override` header is simply being ignored.

If that is the case, Linkerd should be modified to check if the `l5d-dst-override` header is present before using the value of the `Host` header.

**Comment by oncipriani** [comment]
> Are you using the `nginx.ingress.kubernetes.io/service-upstream: "true"` annotation, per https://linkerd.io/2.12/tasks/using-ingress/#nginx? This should eliminate the need to set `l5d-dst-override` headers.

Yes, but it works even without it when using the annotation `nginx.ingress.kubernetes.io/upstream-vhost: $service_name.$namespace.svc.cluster.local`.

I should also mention that our Ingress Controller is **not** meshed, because it uses HostNetwork (we have plans to change this in the future).

> Can you be more detailed about where/how you're looking for these metrics? You can check individual proxy's metrics with `linkerd diagnostics proxy-metrics`, for instance, to spot check which route metrics are present.

Per https://linkerd.io/2.12/tasks/getting-per-route-metrics/#troubleshooting, I am using the following command to check if routes are working:

```sh
linkerd viz -n my-app-namespace tap deploy/my-app-deployment -o wide | grep -Fw rt_route
```

When using the `nginx.ingress.kubernetes.io/upstream-vhost` annotation, the comand above shows the correct metrics in the `rt_route` field. Metrics are also displayed in the "Route Metrics" tab of the Deployment dashboards in Linkerd.

When I remove the `nginx.ingress.kubernetes.io/upstream-vhost` annotation, the command above does not return anything and all metrics appear zeroed in the dashboard.

To summarize:

- When using `nginx.ingress.kubernetes.io/upstream-vhost: $service_name.$namespace.svc.cluster.local` Linkerd **does** collect per route metrics, even **without** `nginx.ingress.kubernetes.io/service-upstream: "true"` and **without** `l5d-dst-override` headers.
- Remove `nginx.ingress.kubernetes.io/upstream-vhost: $service_name.$namespace.svc.cluster.local` and Linkerd **stops** collecting per route metrics, even **with** `nginx.ingress.kubernetes.io/service-upstream: "true"` and `l5d-dst-override` headers.

**Comment by uqix** [comment]
Not-injected ingress/gateway (due to hostNetwork) with `l5d-dst-override` request header added not works, it it true?

**Comment by oncipriani** [comment]
> Not-injected ingress/gateway (due to hostNetwork) with `l5d-dst-override` request header added not works, it it true?

Yes.

**Comment by adleong** [comment]
Hi @oncipriani 

I believe I understand what's going on here.  Typically, when both the source pod (the ingress controller in your case) and the target pod are meshed, the request passes through two linkerd-proxies: first an *outbound* proxy in source pod and then an *inbound* proxy in the target pod.  In your case, however, the ingress controller is not meshed and therefore there is no *outbound* proxy.

The `l5d-dst-override` header controls *outbound* proxy behavior, which is why setting this header is having no effect for you.

When the request reaches in the *inbound* proxy, we look at the `Host` header to determine which service profile (if any) to use for routes.  This is why using `upstream-vhost` results in correct methods: it rewrites the host header to the name of the Service.

Without rewriting the host header in this way or without the presence of an *outbound* proxy (by meshing the ingress controller) there is no way for the *inbound* proxy to look up the service profile and get the correct list of routes.

**Comment by oncipriani** [comment]
Thank you for the explanation, @adleong.

> Without rewriting the host header in this way or without the presence of an _outbound_ proxy (by meshing the ingress controller) there is no way for the _inbound_ proxy to look up the service profile and get the correct list of routes.

Would it be possible to update Linkerd _inbound_ proxies to also look for the `l5d-dst-override` header? For example, checking first if the `l5d-dst-override` is present and, if not, check the `Host` header?

I understand that meshing the Ingress Controller would be the best way of making this work and we have plans to do that, but I believe that the problem would also be solved if the _inbound_ proxy also checked the `l5d-dst-override` header.

This would also have the added benefit of not requiring the `nginx.ingress.kubernetes.io/service-upstream: "true"` annotation, which breaks NGINX's sticky session.

**Comment by adleong** [comment]
Another wrinkle here is that the way this works is currently in flux as we move away from ServiceProfiles and towards the GatewayAPI (see: https://buoyant.io/blog/linkerd-and-the-gateway-api).  HTTPRoute resources are attached to `Server` resources and don't have this issue (unlike `ServiceProfiles` which are associated with `Service` resources).  For more information about the `Server` vs `Service` distinction, see: https://linkerd.io/2.12/reference/authorization-policy/#server
